test/integration_test.py

```diff
@@ -224,50 +223,107 @@ def managed_access_sources_by_catalog(self) -> Dict[CatalogName,
+        partition_prefix = self.random.choice(partition_prefixes)
+        effective_prefix = source.spec.prefix.common + partition_prefix
+        fqids = self.azul_client.list_bundles(catalog, source, partition_prefix)
+        bundle_count = len(fqids)
+        partition = f'Partition {effective_prefix!r} of source {source.spec}'
+        if config.is_main_deployment(config.deployment_stage):
+            # For sources that use partitioning, 512 is the maximum allowed
+            # partition size.
+            upper = 512
+            if effective_prefix:
+                lower = upper // 16
+                if len(fqids) < lower:
+                    # If bundle UUIDs were uniformly distributed by prefix, we
+                    # could directly assert a minimum partition size, but since
+                    # they're not, a given partition may fail to exceed this
+                    # minimum even with an optimal choice of partition prefix
+                    # length.
+                    log.warning('With %i bundles, %s is too small', len(fqids), partition)
+                    counts = plugin.list_partitions(source)
+                    if counts is None:
+                        lower = 1
+                    else:
+                        # For plugins that support efficiently counting bundles
+                        # across all partitions simultaneously, we can check
+                        # whether the chosen partition is an outlier by
+                        # determining the *average* partition size.
+                        bundle_count = sum(counts.values()) / len(counts)
+            else:
+                # Sources too small to be split into more than one partition may
+                # have as few as one bundle in total
+                lower = 1
+        else:
+            # The sandbox and personal deployments typically don't use
+            # partitioning and the desired number of bundles per source is
+            # 16 +/- 50%, i.e. 8 to 24. However, no choice of common prefix can
+            # satisfy this range for snapshots with `n` bundles where `n < 8` or
+            # `24 < n <= 112`. The choice of upper bound here is somewhat
+            # arbitrary.
+            upper = 64
+            lower = 1
+
+        self.assertLessEqual(bundle_count, upper, partition + ' is too large')
+        self.assertGreaterEqual(bundle_count, lower, partition + ' is too small')
+
+        return source, partition_prefix, fqids
 
     def _list_bundles(self,
                       catalog: CatalogName,
                       *,
-                      max_bundles: int,
+                      min_bundles: int,
                       check_all: bool
-                      ) -> List[SourcedBundleFQID]:
-        sources = sorted(self.azul_client.catalog_sources(catalog))
-        # See below why we shuffle
+                      ) -> Iterator[Tuple[SourceRef, str, List[SourcedBundleFQID]]]:
+        total_bundles = 0
```

Comment:
> Might make sense to swap this variable and the comment about shuffling below, pushes the variable close to it's usage.

I think the comment needs to be directly above the for-loop that it describes.

Reply:
> Why the sort if its going to be shuffled?

`Random.shuffle` requires a `Sequence` as an argument and `config.sources` returns a `set`. We use `sorted()` instead of `list()` to ensure that the ordering is deterministic based on the random seed.